### PR TITLE
Gstf patch

### PR DESCRIPTION
**Pull request by anilthanki**

### CODE REVIEW

File: tools/gstf_preparation/gstf_preparation.py [hunk 1]
```diff
@@ -121,6 +125,7 @@ def feature_to_dict(cols, parent_dict=None):
 def add_gene_to_dict(cols, species, gene_dict):
     global gene_count
     gene = feature_to_dict(cols)
+    print(gene)
```
Comment (nsoranzo) [inline tools/gstf_preparation/gstf_preparation.py]: Remove

File: tools/gstf_preparation/gstf_preparation.py [hunk 2]
```diff
@@ -97,6 +98,9 @@ def feature_to_dict(cols, parent_dict=None):
     for attr in cols[8].split(';'):
         if '=' in attr:
             (tag, value) = attr.split('=')
+            if tag == 'biotype':
+                d[tag] = value
+                print("biotype")
```
Comment (nsoranzo) [inline tools/gstf_preparation/gstf_preparation.py]: Remove debugging print

File: tools/gstf_preparation/gstf_preparation.py [hunk 3]
```diff
@@ -247,9 +252,13 @@ def write_gene_dict_to_db(conn, gene_dict):
             for transcript in gene["Transcript"]:
                 transcript_id = transcript['id']
                 protein_id = transcript.get('Translation', {}).get('id', None)
+                if "biotype" in transcript:
+                    biotype = transcript["biotype"]
+                else:
+                    biotype = "NULL"
```
Comment (nsoranzo) [inline tools/gstf_preparation/gstf_preparation.py]: I think `None` should work as well?

File: tools/gstf_preparation/gstf_preparation.py [hunk 4]
```diff
@@ -400,13 +420,33 @@ def __main__():
...
                 break
 
-            gene_transcripts_dict.setdefault(gene_id, []).append((transcript_id, len(entry.sequence)))
+            transcript_type = fetch_biotype_for_transcript(conn, transcript_id)
+            gene_transcripts_dict.setdefault(gene_id, []).append((transcript_id, transcript_type, len(entry.sequence)))
 
     if options.longestCDS:
-        # For each gene, select the transcript with the longest sequence.
-        # If more than one transcripts have the same longest sequence for a
-        # gene, the first one to appear in the FASTA file is selected
```
Comment (nsoranzo) [inline tools/gstf_preparation/gstf_preparation.py:411]: The comment still applies, restore?

**Review by nsoranzo — COMMENTED** [review]
Can you add a test to the tool XML with an example of the problematic GFF3 gene?

File: tools/gstf_preparation/gstf_preparation.py (hunk 4 above)
Comment (anilthanki) [reply tools/gstf_preparation/gstf_preparation.py:411]: yes

File: tools/gstf_preparation/gstf_preparation.py [hunk 5]
```diff
@@ -97,6 +98,8 @@ def feature_to_dict(cols, parent_dict=None):
     for attr in cols[8].split(';'):
         if '=' in attr:
             (tag, value) = attr.split('=')
+            if tag == 'biotype':
+                d[tag] = value
```
Comment (nsoranzo) [inline tools/gstf_preparation/gstf_preparation.py]: These 2 lines are not needed, it's done anyway at (old) line 105.

**Comment by nsoranzo** [comment]
Seems a duplicate at this point.

**Pull request closed without merging**